### PR TITLE
Grab bag of (mostly) typing improvements

### PR DESCRIPTION
**Pull request by benjaminglass1**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #155877
* #158074
* __->__ #158075

Collects some scattershot improvements made while attempting to enable training for AOTInductor. Non-typing changes are:

1. Swapping a few custom searches for the output node in an FX graph for calling `graph.output_node()`.
2. Removing two unused parameters from `torch.export._unlift._unlift`.
3. Switching handles to constants in `cpp_wrapper_cpu` to use C++ references for memory efficiency.
4. Cleaning out unused, unexported imports from `torch/export/__init__.py`, and adding one missing export to `__all__`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

### CODE REVIEW

File: benchmarks/dynamo/common.py [hunk 1]
```diff
@@ -782,7 +783,9 @@ def vary_batch(t: torch.Tensor, new_batch_size) -> torch.Tensor:
     return (time_total, result) if return_result else time_total
```
Comment (Skylion007) [inline benchmarks/dynamo/common.py:772]:
```suggestion
# _d is the exact dict subtype you passed in
_D = TypeVar("_M", bound=dict[str, Any])

@overload
def _normalize_bench_inputs(example_inputs: _D) -> Tuple[Tuple[()], _D]:
    ...

@overload
def _normalize_bench_inputs(example_inputs: Sequence[_T]) -> Tuple[Tuple[_T, ...], Dict[str, Any]]:
    ...
```
Will be far more accurate typing here and preserve the type of the input dictionary and input sequence.

File: torch/_dynamo/utils.py [hunk 2]
```diff
@@ -2115,8 +2115,14 @@ def torch_clone(x):
         return result
 
 
+@overload
+def clone_inputs(example_inputs: dict[str, Any]) -> dict[str, Any]: ...
+@overload
+def clone_inputs(example_inputs: Union[list[Any], tuple[Any]]) -> list[Any]: ...
+
```
Comment (Skylion007) [inline torch/_dynamo/utils.py]:
```suggestion
@overload
def clone_inputs(example_inputs: dict[str, Union[T, tuple[T, ...], list[T]]]) -> dict[str, list[T]]: ...
@overload
def clone_inputs(example_inputs: list[T]) -> list[T]: ...

File: torch/_inductor/compile_fx.py [hunk 3]
```diff
@@ -429,7 +430,7 @@ def _unlift_graph(
 
     from torch.export._unlift import _unlift
 
-    outputs = list(gm.graph.nodes)[-1].args[0]
+    outputs: tuple[torch.fx.Node] = tuple(gm.graph.output_node().args[0])  # type: ignore[arg-type]
```
Comment (Skylion007) [inline torch/_inductor/compile_fx.py]: Is this supposed to be a 1 length tuple??

File: torch/export/exported_program.py [hunk 4]
```diff
@@ -1208,7 +1210,9 @@ def example_inputs(self, value):
     @property
     @compatibility(is_backward_compatible=False)
     def call_spec(self):
-        CallSpec = namedtuple("CallSpec", ["in_spec", "out_spec"])
+        class CallSpec(NamedTuple):
```
Comment (Skylion007) [inline torch/export/exported_program.py:1213]: Thank you for this! <3 Reallly need to replace the rest of these legacy NamedTuples

**Review by Skylion007 — COMMENTED** [review]
Typing nits

File: torch/_inductor/compile_fx.py (hunk 3 above)
Comment (benjaminglass1) [reply torch/_inductor/compile_fx.py]: Ah shoot, good catch, will update.

Comment (Skylion007) [reply torch/_inductor/compile_fx.py]: Other instances of this in the dif that are likely error

**Review by Skylion007 — APPROVED** [review]
Once these nits are resolved. Looks good.

File: torch/export/exported_program.py [hunk 5]
```diff
@@ -610,7 +612,7 @@ def update_arg(old_arg, new_ph):
         raise RuntimeError(f"Type of old_arg not supported: {type(old_arg)}")
 
     new_placeholders = [node for node in gm.graph.nodes if node.op == "placeholder"]
-    new_outputs = list(gm.graph.nodes)[-1].args[0]
+    new_outputs: tuple[torch.fx.Node] = tuple(gm.graph.output_node().args[0])  # type: ignore[arg-type]
```
Comment (Skylion007) [inline torch/export/exported_program.py]: Here too

**Comment by benjaminglass1** [comment]
BC lint is failing on updating some existing types from `tuple[Any]` to `tuple[Any, ...]`. Adding a label to suppress that.

**Comment by benjaminglass1** [comment]
With this amount of test failures, I'm going to back out the changes to `tuple[Any, ...]`; they seem to have broken some non typing things.
